### PR TITLE
Refactor pdb executable targets

### PR DESCRIPTION
**Pull request by gaogaotiantian**
In #26992 the executable target part of pdb is refactored to make code in `Pdb` cleaner. However I often feel like the new code is unnecessarily wordy, especially `_ModuleTarget`. Making `_ModuleTarget` and `_ScriptTarget` a subclass of `str` does not make too much sense to me as they are obviously more than a string. The only possible explanation to me was there's an concatination of strings in the message part.

Even though it removed the separate code path in `Pdb`, it's still not clear what is required to be an executable target. For example, what if I want to add another target (and that's what happened today, so I've decided to fix this first).

It seems like, to be an "executable target", only 3 properties are required - `filename`, `code` and `namespace`. So I had a base class `_ExecutableTarget` with the type hints (it is not public, internal only). For both `_ScriptTarget` and `_ModuleTarget`, we don't need a separate `check` method as that is always needed. The check can be done just in `__init__`. `_ModuleTarget` is significantly simplified by doing `_get_module_details` in `__init__` and just keep the results. It really confuses people (maybe just me) when the class has so many properties and they are referencing each other.

One reason I can imaging for doing that is to make sure everything is "read-only". However, this is an internal data structure that is only used by pdb, and not even in many places. I think we can be responsible enough to not do evil things. I still kept the public facing attributes "properties" as it's cleaner.

This refactor has some very minor observable behavior changes - for example, when the script file does not exist, it will fail when initializing the `_ScriptTarget`, instead of in `Pdb._run()` when it's lazily loaded. I don't think this kind of behavior change need news entry so I skipped both news and issue.

### CODE REVIEW

File: Lib/pdb.py [hunk 1]
```diff
@@ -156,77 +155,75 @@ def __repr__(self):
...
+        return self._target
 
     @property
     def filename(self):
-        return self
+        return self._target
+
+    @property
+    def code(self):
+        # We need to open the file each time because the file may be modified
```
Comment (jaraco) [inline Lib/pdb.py]:
```suggestion
        # Open the file each time because the file may be modified
```

**Review by jaraco — APPROVED** [review]
LGTM. Thanks!

**Comment by gaogaotiantian** [comment]
> I chose a subclass of `str` because that provided a means to give what was previously just a string some more functionality. In that way, the "Target" could simply operate as the target (directory) instead of needing to model both the target and the filename separately. Your approach is also sound I'm glad to see it.

Yeah I think having a somehow unified interface (as a separate class) might help when we add more targets. I actually made this refactor because I want to support zipapps - and realized that it might make the code cleaner if we structure this way. Also having a lot of properties for a `str` subclass is a bit weird for me, but that might just be personal.

**Comment by erlend-aasland** [comment]
FTR, I don't think this change was small enough to earn the `skip-issue` label. Consider creating an issue next time.

**Comment by gaogaotiantian** [comment]
> FTR, I don't think this change was small enough to earn the `skip-issue` label. Consider creating an issue next time.

Sure, could you elaborate a bit on the evaluation? I skipped the issue because
1. It's an internal change that only affects `pdb`
2. It's not observable to users
3. It's a pure refactoring with almost no logic change (except for some implementation details)
4. The code change is not significant

Some of them might be subjective like how much code would make it "significant". Is it because I should explain the motivation to make the refactor in the issue?

**Comment by erlend-aasland** [comment]
I suggest re-reading relevant parts of the devguide, particularly [The life of a pull request](https://devguide.python.org/getting-started/pull-request-lifecycle/#submitting). Our documented and de facto[^1] practice is that pretty much all code changes require an issue. We only skip issue for smaller doc changes, spelling mistakes, and once in a while very minor code changes. For a refactoring, the issue would be a place to explain and discuss the motivation and rationale, as well as connecting possible future PRs (for example bugfixes bco. regressions).


[^1]: our Git log clearly shows that most code changes are connected to an issue

**Comment by gaogaotiantian** [comment]
Got it, thanks! I'll be more careful next time!

**Comment by erlend-aasland** [comment]
Thanks, and thanks for all your good work!

**Pull request merged**